### PR TITLE
Added Roles and Groups Service

### PR DESCRIPTION
**Pull request by thoughtpalette**
- Added Roles and Groups Services to haystack Client
- Added support for using TS generics on Users, Roles and Groups services.

### CODE REVIEW

**Review by shal1y — COMMENTED** [review]
LGTM. I also feel like the user interface should not have an optional user tag. That should be required, hence the user :). I'd also have the id be there.

**Comment by thoughtpalette** [comment]
Updated a with a few more commits, mostly adding Tests and addressing comment(s). 

Feel free to take another look @shal1y ! 🎉

File: src/client/groups/GroupService.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
+import { HDict, HGrid, HMarker, HRef, HaysonDict, HList } from 'haystack-core'
```
Comment (garethj2) [inline src/client/groups/GroupService.ts:5]: Need copyright added.

File: spec/client/roles/RolesService.spec.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2020, J2 Innovations. All Rights Reserved
```
Comment (garethj2) [inline spec/client/roles/RolesService.spec.ts]: Nitpik: copyright is 2021.

File: spec/client/groups/GroupsService.spec.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2020, J2 Innovations. All Rights Reserved
```
Comment (garethj2) [inline spec/client/groups/GroupsService.spec.ts]: Nitpik: copyright is 2021.

File: src/client/groups/GroupService.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import { ClientServiceConfig } from '../ClientServiceConfig'
+import { fetchVal } from '../fetchVal'
+import { dictsToGrid } from '../../util/hval'
+import { encodeQuery } from '../../util/http'
+import { ReadOptions } from '../RecordService'
+
+/**
+ * a group record
```
Comment (garethj2) [inline src/client/groups/GroupService.ts]: Extreme total nitpik: start sentance with a capital letter.

File: src/client/groups/GroupService.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+ */
+export interface Group extends HDict {
+	id?: HRef
+	userGroup: HMarker
+}
+
+/**
+ * optional parameters available to readByFilter query
```
Comment (garethj2) [inline src/client/groups/GroupService.ts]: Extreme total nitpik: start sentance with a capital letter.

File: src/client/groups/GroupService.ts [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+ */
+export type GroupsReadOptions = Omit<ReadOptions, 'unique'>
+
+export class GroupsService<T extends Group = Group> {
```
Comment (garethj2) [inline src/client/groups/GroupService.ts:28]: Should have class level comment describing what this is for.

File: src/client/roles/RolesService.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
+import { HaysonDict, HDict, HGrid, HMarker, HRef } from 'haystack-core'
+import { ClientServiceConfig } from '../ClientServiceConfig'
+import { fetchVal } from '../fetchVal'
+import { dictsToGrid } from '../../util/hval'
+import { encodeQuery } from '../../util/http'
+import { ReadOptions } from '../RecordService'
+
+/**
+ * a role record
```
Comment (garethj2) [inline src/client/roles/RolesService.ts]: Extreme total nitpik: sentances should start with a capital letter.

File: src/client/roles/RolesService.ts [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+ */
+export interface Role extends HDict {
+	id?: HRef
+	userRole: HMarker
+}
+
+/**
+ * optional parameters available to readByFilter query
```
Comment (garethj2) [inline src/client/roles/RolesService.ts]: Extreme total nitpik: sentances should start with a capital letter.

File: src/client/roles/RolesService.ts [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+ */
+export type RolesReadOptions = Omit<ReadOptions, 'unique'>
+
+export class RolesService<T extends Role = Role> {
```
Comment (garethj2) [inline src/client/roles/RolesService.ts:28]: Need to a class level comment.

**Review by garethj2 — CHANGES_REQUESTED** [review]
Looks good part from very minor copyright and comment issues (see comments). It would be nice to add some extra comments for users to understand what these services are and what they're for (consider this library can be used outside of J2).

File: src/client/groups/GroupService.ts (hunk 1 above)
Comment (thoughtpalette) [reply src/client/groups/GroupService.ts:5]: Added.

File: spec/client/roles/RolesService.spec.ts (hunk 2 above)
Comment (thoughtpalette) [reply spec/client/roles/RolesService.spec.ts]: Added.

File: spec/client/groups/GroupsService.spec.ts (hunk 3 above)
Comment (thoughtpalette) [reply spec/client/groups/GroupsService.spec.ts]: Added.

File: src/client/groups/GroupService.ts (hunk 4 above)
Comment (thoughtpalette) [reply src/client/groups/GroupService.ts]: Added.

File: src/client/groups/GroupService.ts (hunk 5 above)
Comment (thoughtpalette) [reply src/client/groups/GroupService.ts]: Added.

File: src/client/groups/GroupService.ts (hunk 6 above)
Comment (thoughtpalette) [reply src/client/groups/GroupService.ts:28]: Added.

File: src/client/roles/RolesService.ts (hunk 7 above)
Comment (thoughtpalette) [reply src/client/roles/RolesService.ts]: Added.

File: src/client/roles/RolesService.ts (hunk 8 above)
Comment (thoughtpalette) [reply src/client/roles/RolesService.ts]: Added.

File: src/client/roles/RolesService.ts (hunk 9 above)
Comment (thoughtpalette) [reply src/client/roles/RolesService.ts:28]: Added.

**Comment by thoughtpalette** [comment]
Should be good for another look @garethj2 🎉

**Pull request merged**